match/lib/match/storage/s3_storage.rb

```diff
@@ -0,0 +1,162 @@
+require_relative '../spaceship_ensure'
+require_relative './interface'
+
+module Match
+  module Storage
+    # Store the code signing identities on AWS S3
+    class S3Storage < Interface
+      attr_reader :s3_bucket
+      attr_reader :s3_region
+      attr_reader :s3_client
+      attr_reader :readonly
+      attr_reader :username
+      attr_reader :team_id
+      attr_reader :team_name
+
+      def self.configure(params)
+        s3_region = params[:s3_region]
+        s3_access_key = params[:s3_access_key]
+        s3_secret_access_key = params[:s3_secret_access_key]
+        s3_bucket = params[:s3_bucket]
+
+        if params[:git_url].to_s.length > 0
+          UI.important("Looks like you still define a `git_url` somewhere, even though")
+          UI.important("you use S3 Storage. You can remove the `git_url`")
+          UI.important("from your Matchfile and Fastfile")
+          UI.message("The above is just a warning, fastlane will continue as usual now...")
+        end
+
+        return self.new(
+          s3_region: s3_region,
+          s3_access_key: s3_access_key,
+          s3_secret_access_key: s3_secret_access_key,
+          s3_bucket: s3_bucket,
+          readonly: params[:readonly],
+          username: params[:username],
+          team_id: params[:team_id],
+          team_name: params[:team_name]
+        )
+      end
+
+      def initialize(s3_region: nil,
+                     s3_access_key: nil,
+                     s3_secret_access_key: nil,
+                     s3_bucket: nil,
+                     readonly: nil,
+                     username: nil,
+                     team_id: nil,
+                     team_name: nil)
+        @s3_bucket = s3_bucket
+        @s3_region = s3_region
+        @s3_client = Fastlane::Helper::S3ClientHelper.new(access_key: s3_access_key, secret_access_key: s3_secret_access_key, region: s3_region)
+        @readonly = readonly
+        @username = username
+        @team_id = team_id
+        @team_name = team_name
+      end
+
+      # To make debugging easier, we have a custom exception here
+      def prefixed_working_directory
+        # We fall back to "*", which means certificates and profiles
+        # from all teams that use this bucket would be installed. This is not ideal, but
+        # unless the user provides a `team_id`, we can't know which one to use
+        # This only happens if `readonly` is activated, and no `team_id` was provided
+        @_folder_prefix ||= currently_used_team_id
+        if @_folder_prefix.nil?
+          # We use a `@_folder_prefix` variable, to keep state between multiple calls of this
+          # method, as the value won't change. This way the warning is only printed once
+          UI.important("Looks like you run `match` in `readonly` mode, and didn't provide a `team_id`. This will still work, however it is recommended to provide a `team_id` in your Appfile or Matchfile")
+          @_folder_prefix = "*"
```

Comment:
Having a folder with a `*` makes me very nervous. Is there a specific reason this has to be `*`?

Reply:
I actually have no idea on this one! This is not something I did so I didn't want to change it 😬 `*` is a bit weird but makes me a nervous to change it 🤔